### PR TITLE
refactor!: clean up `core` exports and synthesizer modules

### LINKED ISSUE #1255: Clean up `core` exports
Currently we export many functions in `@penrose/core` without organizing them in any way. For instance, `compileDomain` is simply re-exported from `@penrose/core/compiler`. Generally, both the top- and module-level APIs need to be cleaned up and standardized for better usability.

### PR DESCRIPTION
**Pull request by wodeni**
# Description

Resolves #1255.

This PR cleans up exported functions and types from `core` and move synthesizer-related modules to `synthesizer-ui`. `@penrose/synthesizer`, the CLI for `core/synthesis` is also deprecated.

# Implementation strategy and design decisions

* We opted for a single entry point for `core` exports, i.e. `dist/index.js`, for all types and functions
* The alternative is [subpath exports](https://nodejs.org/api/packages.html#subpath-exports) but it requires [special configuration on the client side](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-7.html). Therefore, it'll require a manual build system to get something like `import { compileSubstance } from "@penrose/core/compiler/Substance"` to work. 
* `@penrose/synthesizer-ui` is renamed to `@penrose/edgeworth`
* Both `synthesis` and `analysis` folders in `core` are moved to `@penrose/edgeworth`

# Examples with steps to reproduce them

# Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have reviewed any generated registry diagram changes

# Open questions

We might consider banning direct imports of subpaths of `dist` to reduce the surface area of the `core` API.

### CODE REVIEW

File: penrose.code-workspace [hunk 1]
```diff
@@ -48,12 +48,8 @@
...
     },
     {
-      "name": "📦 @penrose/synthesizer",
-      "path": "packages/synthesizer"
-    },
-    {
-      "name": "📦 @penrose/synthesizer-ui",
-      "path": "packages/synthesizer-ui"
+      "name": "📦 @penrose/edgeworth",
+      "path": "packages/edgeworth"
```
Comment (samestep) [inline penrose.code-workspace:52]: Alphabetize pls 🙃

File: package.json [hunk 2]
```diff
@@ -48,8 +48,7 @@
       "packages/examples",
       "packages/optimizer",
       "packages/roger",
-      "packages/synthesizer",
-      "packages/synthesizer-ui",
+      "packages/edgeworth",
```
Comment (samestep) [inline package.json:51]: Alphabetize pls :)

File: packages/edgeworth/tsconfig.json [hunk 3]
```diff
@@ -9,6 +9,7 @@
     "allowSyntheticDefaultImports": true,
     "strict": true,
     "forceConsistentCasingInFileNames": true,
+    "suppressImplicitAnyIndexErrors": true,
```
Comment (samestep) [inline packages/edgeworth/tsconfig.json:12]: What's the dealio here?

**Review by samestep — CHANGES_REQUESTED** [review]
Looks great to me! A couple minor requests; the other important thing is to make sure the tests we've moved to `@penrose/edgeworth` still run after this change.

File: packages/edgeworth/tsconfig.json (hunk 3 above)
Comment (wodeni) [reply packages/edgeworth/tsconfig.json:12]: This is in `core` and `automator` as well so I copied them over. We shouldn't really have this turned on tbh

Comment (wodeni) [reply packages/edgeworth/tsconfig.json:12]: https://github.com/penrose/penrose/issues/1370

**Review by samestep — APPROVED** [review]
Looks great!

**Pull request merged**